core/src/main/scala/io/finch/streaming/StreamFromReader.scala

```diff
@@ -0,0 +1,20 @@
+package io.finch.streaming
+
+import com.twitter.io.{Buf, Reader}
+
+/**
+  * Create stream S[F, Buf] from com.twitter.io.Reader[Buf]
+  */
+trait StreamFromReader[S[_[_], _], F[_]] {
```

Comment:
I'm trying to understand why we need this. Is there a reason, `StreamDecode` can't take `Reader[Buf]` as an input?

Reply:
It's a good question. I spent some time on this structure already and believe it's reasonable.

So the way how modules depend on each other now is following:

- iteratee depends on core and iteratee library
- fs2 depends on core and fs2 library
- circe depends on core, circe-iteratee, and circe-fs2 libraries.

If we remove `StreamReader` frome core it raises some interesting questions.
Who is responsible for transforming `Reader[Buf]` into `S[F[_], Buf]`?
If we merge it with`DecodeStream` and push to JSON modules, they're (only circe now) start concerning about low-level stuff about reading streamed HTTP body.
Also people would have to use `circe` module even if they need only to get simple `Buf` stream out of request. There is no way to default `DecodeStream` for `Buf` as in core we don't know which library is used, and there is now common specific type classes.

Reply:
> Also people would have to use circe module even if they need only to get simple Buf stream out of request. 

Trying to understand this bit in particular. Is this referring to the current state of streaming or the state the PR puts it in? My gut feeling is that we want to be able to stream raw `Buf`s from core and have various libraries utilize that as they need.
Seems to relate to:
> There is no way to default DecodeStream for Buf as in core we don't know which library is used

I'm not understanding this restriction. Feels like we should not need this info in the same way we don't for regular http requests, but I think I'm not understanding something about Streaming as it exists

Reply:
@rpless Sorry if I'm unclear.

In the current "master" status of streaming, one should add `finch-iteratee` dependency to enable endpoints like `enumerateBody` and `enumerateJsonBody` as they don't exist in core module. And circe module also provides decoder for `iteratee.Enumerator`.

In this PR _definition_ of endpoints was moved to core, so methods `streamBody` and `streamJsonBody`. Both of them rely on two type classes: `StreamFromReader` and `StreamDecode`. But due to the fact that core module itself doesn't depend on any streaming library, there is no default implementation of `StreamFromReader` could be provided as there is no such type `S[_[_], _]` available in core. There is default identity defition of `StreamDecode` for `Buf=>Buf` transformation available, though, because the only thing it does is to return `S[_[_], _]` back to caller.
These two type-classes are implemented in additional modules. `StreamFromReader` comes with one of the streaming modules: `fs2` or `iteratee` depending on what you'd like to use.
`StreamDecode` comes from JSON module, because it's the place where we put all the JSON related stuff.

What @vkostyukov proposes is to remove `StreamFromReader` and merge it with `StreamDecode` into single type class. My argument is that it would violate separation of concerns, because JSON library would need to understand all that HTTP `Reader[Buf]` stuff, and users will be forced to use this library as dependency, because it would be the only way to get this type class, even if the only thing you need is to stream bytes.

Reply:
>  My argument is that it would violate separation of concerns, because JSON library would need to understand all that HTTP Reader[Buf] stuff, and users will be forced to use this library as dependency, because it would be the only way to get this type class, even if the only thing you need is to stream bytes.

I think we're mostly on the same page about this. I now see what you mean. Basically, people should be able to `streamBinaryBody[IO, Stream]` w/o specifying the content-type and `DecodeStream`, by definition, should have a notion of content type.

I agree we need two type-classes but I'm not sure I like the name `StreamFromReader`. How about `LiftReader`?

Reply:
To clarify:

 - `streamBinaryBody[F, S]: Endpoint[F, S[F, Buf]]` should not take Content-Type and should only require `LiftReader`
 - `streamJsonBody[F, S, A]: Endpoint[F, S[F, A]]` should require `DecodeStream`.

Reply:
I'm still liking the `LiftReader` name more, but it's not a big deal - we can revisit the names in the later cleanup (these are internal and easy to change).

Reply:
Renamed into `LiftReader`